### DISCUSSION TITLE
Handling output from computational chemistry (or parsing formats with no bonding)

### DISCUSSION THREAD

**Discussion by oliver-s-lee** [rdkit/rdkit#6588] [Q&A]
Hi there, thanks for the nice code and apologies if this is a duplicate, I can see lots of discussion related to this issue but not 100% sure what the intended solution is.

I'm looking to parse the output from computational chemistry programs and load a molecule in RDKit, but I think this problem should apply to any file format that contains atom positions only (no bonding inherently). I have (I think) two choices for loading such a molecule:

1) Use `Chem.MolFromXYZFile()` which natively supports a geometry-only format, but doesn't seem to load the same information as the other MolFromX functions, so if I try to do (for example):
```
mol = Chem.MolFromXYZFile("MyFile.xyz")
rank = Chem.rdmolfiles.CanonicalRankAtoms(mol)
```
I get:
```****
Pre-condition Violation
getNumImplicitHs() called without preceding call to calcImplicitValence()
Violation occurred on line 297 in file /project/build/temp.linux-x86_64-cpython-310/rdkit/Code/GraphMol/Atom.cpp
Failed Expression: d_implicitValence > -1
****
```
And the lack of bonding means images aren't much use either:

![SeMol skeletal](https://github.com/rdkit/rdkit/assets/9111160/16226696-c826-4d22-a82f-51f9303bf6db)

From some googling I can fix these issues by calling :
```
mol.UpdatePropertyCache()
rdkit.Chem.rdDetermineBonds.DetermineConnectivity(mol)
```
Now `CanonicalRankAtoms()` returns something and I can render images, but there's no bond order which is a bit sad:

![SeMol skeletal](https://github.com/rdkit/rdkit/assets/9111160/0d6abc17-ed33-46f7-be93-42387577337a)

`rdDetermineBonds.DetermineBonds()` would fix this I believe, but it says: `determineBondOrdering() does not work with element Se`

2) First convert the xyz to mol with obabel, then load with `MolFromMolFile()`. This works great for trivial structures, but for 'complex' ones with eg dative bonds:
```
[17:35:57] Explicit valence for atom # 11 B, 4, is greater than permitted
```
This can be solved/ignored by setting `sanitize = False`, but I'm not sure if this is generally considered safe? Doing so gets some bond orders, which is cool, but they're of questionable accuracy. There's also some funky stereochemistry now, in this example at least:

![SeMol skeletal](https://github.com/rdkit/rdkit/assets/9111160/f2ea92ab-4d4a-47a4-9274-e691aa3557a8)

My question essentially is what is the recommended way of dealing with this scenario? Is there a better way of parsing geometry only files?

**Comment by ptosco** [comment]
Your best option if you have a topology definition for your molecules is to use
[AllChem.AssignBondOrdersFromTemplate](https://www.rdkit.org/docs/source/rdkit.Chem.AllChem.html?highlight=allchem#rdkit.Chem.AllChem.AssignBondOrdersFromTemplate), as it will avoid having to guess bond orders from 3D coordinates. The latter is a non-trivial and often underdetermined task, which may yield undesired results.

**Comment by greglandrum** [comment]
@oliver-s-lee, as @ptosco said: this is a non-trivial topic.
If you know the net charges on the molecules and the XYZ files have Hs in them (both of these are usually true when working with data from QM calculations), then you might try using the functionality in `rdDetermineBonds`: https://greglandrum.github.io/rdkit-blog/posts/2022-12-18-introducing-rdDetermineBonds.html

**Comment by oliver-s-lee** [comment]
Hi both, thanks for your suggestions. `rdDetermineBonds` is what I'm using now and seems to work well for simple organics, but fails for more 'exotic' atoms (such as Se), which is a bit of shame, but I understand there is not necessarily a simple solution to this problem.

Conversely, the obabel -> mol -> `MolFromMolFile()` route seems to have better support for different atom types, but struggles with more 'exotic' bonding (dative bonds, for example). I'm not sure if this is because obabel saves the 'strange' bonds incorrectly or because rdkit cannot parse them, I'm not particularly familiar with the mol format.

For completeness Obabel renders the above structure as this:
 ![SeMol](https://github.com/rdkit/rdkit/assets/9111160/f185bd2c-429f-4c7d-9499-d598ff9005f3)
which better indicates the conjugation than rdkit but conversely suggests the molecule is both a radical and non-planar, which isn't great either. Maybe there's some avenue for a joint solution, but I'm not sure what that would be.

**Comment by ghutchis** [comment]
> fails for more 'exotic' atoms (such as Se)

This seems strange to me, since selenocysteine is a thing. Can Se be added as a larger S atom? I would think for most cases, this would solve a range of problems.

The `obabel` output is weird. The correct structure, naturally would have [B-] and then a double bond to the nitrogen to maintain planarity and generate [N+] balancing the charge.

I've been looking at the `xyz2mol` algorithm implemented in RDKit and I'm trying to see how it compares to the ancient method in Babel (and a random forest model).
